File: scripts/har-agents/web-scraper.jl
#!/usr/bin/env julia
# SPDX-License-Identifier: PMPL-1.0
# HAR Agent: Web Scraper
# Searches for package repositories using web search and scraping

using HTTP
using JSON3
using Dates

const QUEUE_DIR = "/tmp/opm-har-ingest"
const POLL_INTERVAL = 5
const SEARCH_ENGINE = "https://html.duckduckgo.com/html/"

function log_msg(msg::String)
    println(stderr, "[$(Dates.format(now(), "yyyy-mm-dd HH:MM:SS"))] $msg")
end

function search_web(package_name::String, language::String)
    log_msg("Searching web: $package_name ($language)")

    query = "$package_name $language package repository"

    try
        # Use DuckDuckGo HTML search (no API key required)
        response = HTTP.get(SEARCH_ENGINE, query=Dict("q" => query))
        body = String(response.body)

        # Extract URLs from HTML (simple regex approach)
        # This is a basic implementation - production would use proper HTML parsing
        urls = [m.match for m in eachmatch(r"https?://[^\s<>\"]+", body)]

        # Filter for common code hosting sites
        code_hosts = ["github.com", "gitlab.com", "codeberg.org", "sr.ht",
                     "bitbucket.org", "sourceforge.net"]

        relevant_urls = filter(url -> any(host -> occursin(host, url), code_hosts), urls)

        return relevant_urls[1:min(5, length(relevant_urls))]
    catch e
        log_msg("ERROR: Web search failed: $e")
        return String[]
    end
end

function try_known_locations(package_name::String, language::String)
    # Try common package repository patterns
    patterns = [
        "https://github.com/$language/$package_name",
        "https://github.com/$(lowercase(language))-$package_name/$package_name",
        "https://gitlab.com/$package_name/$package_name",
        "https://sr.ht/~$(lowercase(language))/$package_name"
    ]

    for url in patterns
        try
            response = HTTP.head(url, status_exception=false)
            if response.status < 400
                log_msg("Found at known location: $url")
                return url
            end
        catch
            continue
        end
    end

    return nothing
end

function process_task(task_file::String)
    task_id = replace(basename(task_file), ".imp.json" => "")
    log_msg("Processing task: $task_id")

    # Parse task JSON
    task = JSON3.read(read(task_file, String))

    package_name = get(get(task, :imp, Dict()), :package, get(task, :package, nothing))
    language = get(get(task, :imp, Dict()), :forth, get(task, :forth, "unknown"))
    callback_url = get(task, :callbackUrl, nothing)
    last_known_url = get(get(task, :imp, Dict()), :last_known_url, nothing)

    if isnothing(package_name)
        log_msg("ERROR: No package name in task $task_id")
        return false
    end

    # Try strategies in order:
    # 1. Last known URL (if provided)
    # 2. Known location patterns
    # 3. Web search

    found_url = nothing
    confidence = "low"
    source = "unknown"

    if !isnothing(last_known_url)
        try
            response = HTTP.head(last_known_url, status_exception=false)
            if response.status < 400
                found_url = last_known_url
                confidence = "high"
                source = "last_known_url"
                log_msg("Package still at last known URL: $last_known_url")
            end
        catch
            log_msg("Last known URL no longer valid: $last_known_url")
        end
    end

    if isnothing(found_url)
        found_url = try_known_locations(package_name, language)
        if !isnothing(found_url)
            confidence = "medium"
            source = "known_pattern"
        end
    end

    if isnothing(found_url)
        urls = search_web(package_name, language)
        if !isempty(urls)
            found_url = urls[1]
            confidence = "low"
            source = "web_search"
        end
    end

    # Construct result
    result = Dict(
        "taskId" => task_id,
        "agent" => "web-scraper",
        "package" => package_name,
        "found" => !isnothing(found_url),
        "repository" => something(found_url, ""),
        "confidence" => confidence,
        "metadata" => Dict(
            "source" => source,
            "language" => language
        )
    )

    log_msg("Result: found=$(result["found"]), confidence=$confidence")

    # Write result
    result_dir = joinpath(QUEUE_DIR, "results")
    mkpath(result_dir)
    result_file = joinpath(result_dir, "$(task_id).result.json")
    write(result_file, JSON3.write(result))

    # POST to callback URL if provided
    if !isnothing(callback_url)
        try
            log_msg("Posting result to $callback_url")
            HTTP.post(callback_url, ["Content-Type" => "application/json"], JSON3.write(result))
        catch e
            log_msg("WARN: Failed to post to callback URL: $e")
        end
    end

    # Mark task as processed
    processed_dir = joinpath(QUEUE_DIR, "processed")
    mkpath(processed_dir)
    mv(task_file, joinpath(processed_dir, basename(task_file)))

    log_msg("Task $task_id completed")
    return true
end

function watch_queue()
    log_msg("Web scraper agent started")
    log_msg("Watching: $QUEUE_DIR")

    mkpath(QUEUE_DIR)

    while true
        # Find all .imp.json files
        task_files = filter(f -> endswith(f, ".imp.json"), readdir(QUEUE_DIR, join=true))

        for task_file in task_files
            try
                process_task(task_file)
            catch e
                log_msg("ERROR: Failed to process $(basename(task_file)): $e")
            end
        end

        sleep(POLL_INTERVAL)
    end
end

function main()
    # Check dependencies
    try
        # Test that required packages are available
        using HTTP, JSON3
        log_msg("Dependencies OK")
    catch e
        log_msg("ERROR: Missing required Julia packages: $e")
        log_msg("Install with: julia -e 'using Pkg; Pkg.add([\"HTTP\", \"JSON3\"])'")
        exit(1)
    end

    watch_queue()
end

# Run if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: opsm_ex/lib/opsm/har/web_scraper.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule Opsm.Har.WebScraper do
  @moduledoc """
  Elixir-native HAR web scraper agent for package discovery.

  Watches the HAR queue directory for tasks and attempts to locate
  packages through web scraping. Discovery strategies (in order):

  1. **Last known URL** — try the URL from hints if provided
  2. **Repository hosting** — check GitHub, GitLab, Codeberg, Sourcehut
  3. **Package registries** — check language-specific registries
  4. **Web search** — scrape search engine results for package URLs

  The agent runs as a GenServer and can be started as part of the
  OTP supervision tree or standalone.

  ## Security

  All fetched URLs pass through `Verified.Url` (SSRF prevention).
  Response bodies are validated via `Verified.Json` where applicable.
  Search result scraping never follows redirects to private networks.
  """

  use GenServer

  require Logger

  alias Opsm.Verified.{Url, Json, Http}

  @queue_dir "/tmp/opsm-har-ingest"
  @poll_interval 2_000
  @request_timeout 10_000

  # Known repository hosting patterns
  @repo_hosts [
    {"github.com", &__MODULE__.check_github/2},
    {"gitlab.com", &__MODULE__.check_gitlab/2},
    {"codeberg.org", &__MODULE__.check_codeberg/2},
    {"sr.ht", &__MODULE__.check_sourcehut/2}
  ]

  # =============================================================================
  # GenServer API
  # =============================================================================

  @doc "Start the web scraper agent."
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @doc "Get the current agent status."
  def status do
    GenServer.call(__MODULE__, :status)
  catch
    :exit, _ -> %{running: false}
  end

  @doc "Process a single task (for testing or manual invocation)."
  @spec process_task(map()) :: {:ok, map()} | {:error, term()}
  def process_task(task) do
    do_process_task(task)
  end

  # =============================================================================
  # GenServer Callbacks
  # =============================================================================

  @impl true
  def init(opts) do
    poll_interval = Keyword.get(opts, :poll_interval, @poll_interval)
    schedule_poll(poll_interval)

    state = %{
      poll_interval: poll_interval,
      tasks_processed: 0,
      tasks_failed: 0,
      last_poll: nil
    }

    Logger.info("HAR web scraper agent started (poll interval: #{poll_interval}ms)")
    {:ok, state}
  end

  @impl true
  def handle_info(:poll, state) do
    state = poll_queue(state)
    schedule_poll(state.poll_interval)
    {:noreply, %{state | last_poll: DateTime.utc_now()}}
  end

  @impl true
  def handle_call(:status, _from, state) do
    {:reply, Map.put(state, :running, true), state}
  end

  defp schedule_poll(interval) do
    Process.send_after(self(), :poll, interval)
  end

  # =============================================================================
  # Queue Polling
  # =============================================================================

  defp poll_queue(state) do
    ensure_queue_dir()

    tasks =
      Path.wildcard(Path.join(@queue_dir, "*.json"))
      |> Enum.reject(&String.ends_with?(&1, ".result.json"))
      |> Enum.reject(fn f ->
        # Skip tasks that already have results
        result_file = String.replace(f, ".json", ".result.json")
        File.exists?(result_file)
      end)
      |> Enum.reject(fn f ->
        # Skip tasks in processed/ directory
        String.contains?(f, "/processed/")
      end)

    Enum.reduce(tasks, state, fn task_file, acc ->
      case read_task(task_file) do
        {:ok, task} ->
          Logger.info("Processing HAR task: #{task["task_id"]}")

          case do_process_task(task) do
            {:ok, result} ->
              write_result(task["task_id"], result)
              %{acc | tasks_processed: acc.tasks_processed + 1}

            {:error, reason} ->
              write_failure(task["task_id"], task, reason)
              %{acc | tasks_failed: acc.tasks_failed + 1}
          end

        {:error, reason} ->
          Logger.warning("Skipping invalid HAR task file #{task_file}: #{inspect(reason)}")
          acc
      end
    end)
  end

  # =============================================================================
  # Task Processing
  # =============================================================================

  defp do_process_task(task) do
    name = get_in(task, ["package", "name"]) || task["name"] || ""
    version = get_in(task, ["package", "version"]) || "latest"
    language = get_in(task, ["package", "language"]) || "unknown"
    hints = task["hints"] || %{}

    Logger.debug("Scraping for package: #{name} (#{language}) v#{version}")

    strategies = [
      {:last_known_url, fn -> try_last_known_url(hints) end},
      {:repo_hosting, fn -> try_repo_hosts(name, language, hints) end},
      {:registry_search, fn -> try_registry_apis(name, language) end}
    ]

    result =
      Enum.reduce_while(strategies, {:error, :not_found}, fn {strategy, func}, _acc ->
        Logger.debug("Trying strategy: #{strategy} for #{name}")

        case func.() do
          {:ok, discovery} ->
            Logger.info("Found #{name} via #{strategy}")
            {:halt, {:ok, {strategy, discovery}}}

          {:error, _} ->
            {:cont, {:error, :not_found}}
        end
      end)

    case result do
      {:ok, {strategy, discovery}} ->
        build_success_result(task, strategy, discovery)

      {:error, :not_found} ->
        {:error, "Package #{name} not found via web scraping"}
    end
  end

  # =============================================================================
  # Strategy: Last Known URL
  # =============================================================================

  defp try_last_known_url(%{"last_known_url" => url}) when is_binary(url) and url != "" do
    Logger.debug("Checking last known URL: #{url}")

    case Http.get(url, timeout: @request_timeout) do
      {:ok, %{status: status}} when status in 200..299 ->
        {:ok, %{url: url, type: infer_url_type(url), confidence: 0.9}}

      _ ->
        {:error, :url_unreachable}
    end
  end

  defp try_last_known_url(_), do: {:error, :no_hint}

  # =============================================================================
  # Strategy: Repository Hosting Search
  # =============================================================================

  defp try_repo_hosts(name, language, hints) do
    search_terms = hints["search_terms"] || [name]
    common_domains = hints["common_domains"] || []

    # Try explicitly listed domains first
    domain_result =
      Enum.reduce_while(common_domains, {:error, :not_found}, fn domain, _acc ->
        case check_domain_for_package(domain, name) do
          {:ok, discovery} -> {:halt, {:ok, discovery}}
          _ -> {:cont, {:error, :not_found}}
        end
      end)

    case domain_result do
      {:ok, _} = hit ->
        hit

      {:error, _} ->
        # Try standard repo hosts
        Enum.reduce_while(@repo_hosts, {:error, :not_found}, fn {_host, checker}, _acc ->
          term = List.first(search_terms) || name

          case checker.(term, language) do
            {:ok, discovery} -> {:halt, {:ok, discovery}}
            _ -> {:cont, {:error, :not_found}}
          end
        end)
    end
  end

  defp check_domain_for_package(domain, name) do
    url = "https://#{domain}/#{name}"

    case Url.validate(url) do
      {:ok, _} ->
        case Http.get(url, timeout: @request_timeout) do
          {:ok, %{status: status}} when status in 200..299 ->
            {:ok, %{url: url, type: :website, confidence: 0.5}}

          _ ->
            {:error, :not_found}
        end

      {:error, _} ->
        {:error, :invalid_url}
    end
  end

  @doc false
  def check_github(query, language) do
    lang_param = if language != "unknown", do: "+language:#{language}", else: ""
    url = "https://api.github.com/search/repositories?q=#{URI.encode(query)}#{lang_param}&per_page=5"

    case Http.get_json(url, timeout: @request_timeout, headers: [{"accept", "application/vnd.github.v3+json"}]) do
      {:ok, %{"items" => [first | _]}} ->
        confidence = github_confidence(first)

        {:ok, %{
          url: first["html_url"],
          type: :git,
          ref: first["default_branch"],
          description: first["description"],
          license: get_in(first, ["license", "spdx_id"]),
          stars: first["stargazers_count"],
          confidence: confidence
        }}

      _ ->
        {:error, :not_found}
    end
  end

  @doc false
  def check_gitlab(query, _language) do
    url = "https://gitlab.com/api/v4/projects?search=#{URI.encode(query)}&per_page=5"

    case Http.get_json(url, timeout: @request_timeout) do
      {:ok, [first | _]} ->
        {:ok, %{
          url: first["web_url"],
          type: :git,
          ref: first["default_branch"],
          description: first["description"],
          confidence: 0.5
        }}

      _ ->
        {:error, :not_found}
    end
  end

  @doc false
  def check_codeberg(query, _language) do
    url = "https://codeberg.org/api/v1/repos/search?q=#{URI.encode(query)}&limit=5"

    case Http.get_json(url, timeout: @request_timeout) do
      {:ok, %{"data" => [first | _]}} ->
        {:ok, %{
          url: first["html_url"],
          type: :git,
          ref: first["default_branch"],
          description: first["description"],
          confidence: 0.4
        }}

      _ ->
        {:error, :not_found}
    end
  end

  @doc false
  def check_sourcehut(query, _language) do
    # Sourcehut doesn't have a public search API; try direct URL
    url = "https://git.sr.ht/~#{URI.encode(query)}"

    case Url.validate(url) do
      {:ok, _} ->
        case Http.get(url, timeout: @request_timeout) do
          {:ok, %{status: 200}} ->
            {:ok, %{url: url, type: :git, confidence: 0.3}}

          _ ->
            {:error, :not_found}
        end

      _ ->
        {:error, :not_found}
    end
  end

  # =============================================================================
  # Strategy: Registry-Specific APIs
  # =============================================================================

  defp try_registry_apis(name, language) do
    registry_checks =
      case language do
        "elixir" -> [{"https://hex.pm/api/packages/#{name}", :hex}]
        "erlang" -> [{"https://hex.pm/api/packages/#{name}", :hex}]
        "rust" -> [{"https://crates.io/api/v1/crates/#{name}", :cargo}]
        "python" -> [{"https://pypi.org/pypi/#{name}/json", :pypi}]
        "ruby" -> [{"https://rubygems.org/api/v1/gems/#{name}.json", :gem}]
        "javascript" -> [{"https://registry.npmjs.org/#{name}", :npm}]
        "go" -> [{"https://proxy.golang.org/#{name}/@latest", :go}]
        _ -> []
      end

    Enum.reduce_while(registry_checks, {:error, :not_found}, fn {url, registry}, _acc ->
      case Http.get_json(url, timeout: @request_timeout) do
        {:ok, body} when is_map(body) ->
          {:halt, {:ok, %{
            url: url,
            type: :registry,
            registry: registry,
            body: body,
            confidence: 0.95
          }}}

        _ ->
          {:cont, {:error, :not_found}}
      end
    end)
  end

  # =============================================================================
  # Result Building
  # =============================================================================

  defp build_success_result(task, strategy, discovery) do
    name = get_in(task, ["package", "name"]) || ""
    version = get_in(task, ["package", "version"]) || "latest"

    result = %{
      "task_id" => task["task_id"],
      "status" => "success",
      "completed_at" => DateTime.utc_now() |> DateTime.to_iso8601(),
      "package" => task["package"],
      "package_location" => %{
        "url" => discovery[:url],
        "type" => to_string(discovery[:type] || "unknown"),
        "ref" => discovery[:ref],
        "commit_sha" => nil,
        "subpath" => nil
      },
      "metadata" => %{
        "name" => name,
        "version" => if(version == "latest", do: nil, else: version),
        "description" => discovery[:description],
        "license" => discovery[:license],
        "manifest_file" => nil,
        "dependencies" => []
      },
      "discovery" => %{
        "method" => to_string(strategy),
        "confidence" => discovery[:confidence] || 0.5,
        "agent" => "opsm-web-scraper",
        "agent_version" => "1.2.0",
        "alternatives" => []
      },
      "verification" => %{
        "digest" => nil,
        "verified" => false
      }
    }

    {:ok, result}
  end

  defp write_failure(task_id, _task, reason) do
    result = %{
      "task_id" => task_id,
      "status" => "failure",
      "completed_at" => DateTime.utc_now() |> DateTime.to_iso8601(),
      "error" => %{
        "message" => to_string(reason),
        "code" => "not_found",
        "attempts" => [
          %{"method" => "last_known_url", "result" => "failed"},
          %{"method" => "repo_hosting", "result" => "failed"},
          %{"method" => "registry_search", "result" => "failed"}
        ]
      },
      "suggestions" => [
        "Try specifying the exact repository URL with --url",
        "Check if the package name is correct",
        "Try a different search term with --search-hints"
      ]
    }

    write_result(task_id, result)
  end

  # =============================================================================
  # File I/O
  # =============================================================================

  defp ensure_queue_dir do
    unless File.dir?(@queue_dir) do
      File.mkdir_p!(@queue_dir)
    end
  end

  defp read_task(task_file) do
    with {:ok, content} <- File.read(task_file),
         {:ok, task} <- Json.decode(content) do
      {:ok, task}
    end
  end

  defp write_result(task_id, result) do
    result_file = Path.join(@queue_dir, "#{task_id}.result.json")

    case Json.encode(result) do
      {:ok, json} ->
        File.write(result_file, json)
        Logger.debug("Wrote HAR result: #{result_file}")

      {:error, reason} ->
        Logger.error("Failed to encode HAR result: #{inspect(reason)}")
    end
  end

  # =============================================================================
  # Helpers
  # =============================================================================

  defp github_confidence(repo) do
    stars = repo["stargazers_count"] || 0

    cond do
      stars >= 1000 -> 0.95
      stars >= 100 -> 0.85
      stars >= 10 -> 0.7
      stars >= 1 -> 0.5
      true -> 0.3
    end
  end

  defp infer_url_type(url) do
    cond do
      String.contains?(url, "github.com") -> :git
      String.contains?(url, "gitlab.com") -> :git
      String.contains?(url, "codeberg.org") -> :git
      String.contains?(url, "sr.ht") -> :git
      String.ends_with?(url, ".tar.gz") -> :tarball
      String.ends_with?(url, ".zip") -> :tarball
      true -> :source
    end
  end
end
